[#1616] allow classifiers for CopyDependencies.

bmarwell opened the pull request:
fixes #1616 , fixes #1585

Sadly, @cherylking closed one of the issues.
I want to take this chance to show how small of a fix this would be.

Also, this is VERY convenient, because it is really difficult to let the `dependency-maven-plugin` execute at the right time.

Furthermore, these methods are used all over the place and certainly this will be helpful at some point!

cherylking commented:
@bmarwell First of all, thank you for the PR. This is exactly why we moved these plugins to open source. So that developers like you can contribute to the tools you use and make them better!

> Sadly, @cherylking closed one of the issues.
> I want to take this chance to show how small of a fix this would be.

Second of all, there was a legitimate alternative listed in #1585 that allows this scenario to work. My team is responsible for a lot of different tools and products and we have to choose which suggestions to work on based on the demand/need for something, not based on how small of a fix it is. We had one request for this, and the alternative was acceptable to that user. So why wouldn't I close the issue? 

Finally, in order to accept this PR, it will need a test case demonstrating that the new functionality works and also does not regress anything. The CLA will also need to be signed. See the instructions listed in the [CONTRIBUTING](https://github.com/OpenLiberty/ci.maven/blob/main/CONTRIBUTING.md) doc.

cherylking commented on liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/AbstractLibertySupport.java:
```diff
@@ -251,16 +257,18 @@ protected ArtifactItem createArtifactItem(String groupId, String artifactId, Str
         if (version != null) {
             // if version is set, it will always override the one in project dependency
-            Artifact artifact = getArtifact(groupId, artifactId, type, version);
+            Artifact artifact = getArtifact(groupId, artifactId, type, version, classifier);
```
So is the `classifier` only going to be supported when the `version` is specified? And no wildcards for groupId and artifactId?

cherylking replied:
Also, need to include the classifier in the log warning at line 276.

cherylking replied:
At line 321 do we need to include the classifier?

bmarwell replied:
Added to the logging message.
About your question: 

> So is the classifier only going to be supported when the version is specified? 

Good. If no version is specified, it will try to use the dependencies from the main project, which may ambiguous (depending on how you want to treat classifiers)9. No one will prevent you from  adding the same dependency with different versions and different classifiers to a maven project. Each `g:a:v:c` can be unique. If you only have `g:a`, what will you choose? The only option can only be the dependency where `c == null` (i.e. not set).

If you only have `g:a::c` (no version), the only correct way should be to search for a coordinate with the same `g:a::c` in your dependencies. Will add that in a later commit.

It needs to be included inside THIS monster of a diff (which I surely will refactor out into a method and write a unit test for…):

```java
                if (artifact.getGroupId().equals(groupId) && 
                    ((compareArtifactId == null) ||
                     (isWildcard && artifact.getArtifactId().startsWith(compareArtifactId)) ||
                     (artifact.getArtifactId().equals(compareArtifactId)))) {
```

… although I see no unit tests. Is this intentional? 🤔 

Line 321: Yes, if I get this correctly, this is the line where no version was specified?

cherylking replied:
Yes, no unit tests are intentional. We run integration tests for every build and test functionality the way the user would use it.

If we only want to honor the classifier when the version is specified, I think a check needs to be added somewhere. I'm not sure after your examples where you landed on that point. 

btw..we do have some existing ITs for copyDependencies...in `appsdirectory-notset-configured-it/pom.xml` and `appsdirectory-notset-notconfigured-it/pom.xml`.

bmarwell commented:
> Finally, in order to accept this PR, it will need a test case demonstrating that the new functionality works and also does not regress anything. The CLA will also need to be signed. See the instructions listed in the [CONTRIBUTING](https://github.com/OpenLiberty/ci.maven/blob/main/CONTRIBUTING.md) doc.

I would add unit tests, but I saw no unit tests and thought... well... 🤷🏻‍♂️ 
Would an IT do for you? Those are in `src/it`. Can add one or two.

CLA signed a few years ago and renewed this year b/c of GitHub username change ✅

cherylking commented:
> Would an IT do for you? Those are in src/it. Can add one or two.

Yes, and integration test is what is needed. Thanks. I am adding some comments to the PR review right now.

cherylking commented on liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/StartDebugMojoSupport.java:
```diff
@@ -462,7 +462,7 @@ private void copyDependencies(Dependency dep, String overrideLocation, String de
                     targetFileName = stripVersionFromName(targetFileName, nextArtifact.getVersion());
                 }
-           
+
```
Now that we are taking a `classifier`, do we need to provide `stripClassifier` similar to `stripVersion`?

bmarwell replied:
Even worse, we should make sure (using a unit test)9 that `stripVersionFromName` will still work correctly!

As for the `stripClassifier`: It is uncommon to include a dependency with multiple versions (into the same directory). The same cannot be said about artifacts with classifiers. For this reason, we should not implement this.

That said, looking into the future, `getFile().getName()` and then operating on it might not be a wise choice: Johnzon's jakarta dependency is `apache-johnzon-1.20.4-jakarta.jar` and you could also have artifacts like `my-webapp-1.0.0-static.tar.gz` and `my-webapp-1.0.0-properties.tar.gz` next to `my-webapp-1.0.0.war`. Stripping the qualifiers would make you loose a file.

Therefore, if you want to have this option in this commit, a rewrite to manually concatinating the gid-artifactid-version-classifier.type would be feasible here.

cherylking replied:
And this is why we didn't implement this small quick fix when the original issue was opened. There are other factors to consider, more design to think of. Unfortunately I cannot spend more time on this right now as I have other pressing deadlines.

I will look at your additions/changes as time permits. Thank you for spending time to add doc and tests.

bmarwell replied:
> There are other factors to consider, more design to think of. 

Yes, I find the current design fragile. You are currently relying on the file naming of the file in the maven Repository. Instead, if you added the components manually (e.g. join artifactId, version, type) you would not need a complex stripVersion method and had a stable method.

bmarwell commented on liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/AbstractLibertySupport.java:
```diff
@@ -221,22 +221,28 @@ protected Artifact getResolvedArtifact(final ArtifactItem item) throws MojoExecu
      *             Failed to create artifact
      */
-    protected Artifact getArtifact(String groupId, String artifactId, String type, String version) throws MojoExecutionException {
+    protected Artifact getArtifact( String groupId, String artifactId, String type, String version, String classifier ) throws MojoExecutionException {
```
Thought about it and ditched the idea (for whatever reason I don't remember), but will do.

bmarwell commented:
About updating the readme: Your MD file has `broken` tables according to IntelliJ. This is what they *should* look like (shudder). Since this becomes unreadable in text with line breaks, I adopted the existing style.

![2022-12-08T214356_screenshot](https://user-images.githubusercontent.com/1413391/206564533-4734f050-07db-4e87-aa3f-5a303ae5daed.png)

bmarwell commented:
Hi Cheryl. I refactored the big if condition into a method to take into account the classifiers more easily. I think it is more readable. Also, I added a few unit tests because they are much easier to understand and to test for this logic. Feel free to remove them and to convert them to integration tests if you feel like it.

I also implemented wildcard matching for classifiers.

mattbsox reviewed (changes requested):
Changes look good to me. One last thing to add would be updated copyright dates for 2023 where applicable.

mattbsox reviewed (approved):
Looks good. Thanks!
